#91 - Only consume messages if a worker is available, added listener unit tests, removed redis as a broker

xeonneon opened the pull request:
This PR closes #91

## Introduced Changes
- Removed the `Filesystem` broker from the `runner` because it does not support monitoring, is poorly documented, and does not appear to have first-class support like Redis or RabbitMQ
- Listener will not consumer another message from the queue unless a worker is available to consume the message
  - A worker is available if one of its spawned sub-processes is not currently executing a task. This can be determined by comparing the tasks assigned to the worker with the worker's concurrency setting. If the number of assigned tasks is less than the worker's number of concurrent processes, then the worker has an available process to consume another message from the queue.
- Added unit tests to the listener to test the functionality of the busy/wait logic
- Removed Redis from Functionary in favor of just using RabbitMQ as the message broker
- Added `pytest` and `pytest-mock` to the `requirements-dev.txt` for the `runner`

## Testing
- To test the added changes to the `runner.listener`, run the unit tests found in the `runner/tests/`
  - Open to suggestions on how I should change the structure of the unit tests
- Start Functionary, and then inspect any of the logs for containers that have a Celery worker inside of them
  - You should see 3 total workers join a party, since they are all connecting to the same message broker
  - Publish a package to Functionary. You should still see the proper worker consume the task and build the package
  - Run a Functionary task. You should still see the proper worker consume the task and publish the results


## Celery Worker isolation decision that will need to be made
Since all of the Celery workers are connecting to the same messaging broker and `vhost`, they are all visible to each other, hence why the Celery logs show that another worker joined the party. If we want to separate out the visibility of the Celery workers from each other, I believe we could use [RabbitMQ's `vhost` setting](https://www.rabbitmq.com/vhosts.html), such that the `runner` can connect to one or many separate `vhost`s, and let the functionary and functionary builder workers live on a separate `vhost`. I believe this would prevent the Celery workers from being visible to one another, unless they are connected to the same `vhost`. I think this could also help us when we want to implement private `runner`s, since we could just spawn another `vhost` for that specific runner via RabbitMQ's API.

xeonneon commented on runner/tests/conftest.py:
```diff
@@ -0,0 +1,20 @@
+]
+
+
+@pytest.fixture(scope="session")
```
I put that there for future tests that may need a celery task be run. I can remove them for this PR if you want, but I suspect we will eventually need to write unit tests for the rest of the runner.

scott-taubman replied:
Yeah, I would remove it for now.  Until we actually have tests that use the things defined here, it seems like to just be a source of confusion regarding what it's intended to do.

xeonneon commented:
Working with the idea that each worker is isolated to its own vhost, the actual logic for determining if the worker is busy does not really change other than specifically looking at the first index of the lists returned from Celery's `Inspect` tool or using the nodes hostname as the key lookup. I can make that change, but it doesn't really seem like clean code or necessary anymore.

Looking at the difference between `basic_consume` vs `basic_get`,  I believe we still want to have a consumer, but we just need to limit the consumption of unack'd messages in a channel. This can be done by setting the [`qos` for the channel](https://www.rabbitmq.com/consumer-prefetch.html). This prevents the listener from consuming more unack'd messages than the `qos` limit. Currently, I am setting the `qos` of the channel that the consumer binds to equal to `cpu_count()`, which currently matches the number of concurrent processes I am allowing the Celery worker to spawn. Looking at the RabbitMQ UI, this seems to accomplish the end goal. My only concern with this approach is that if we decide we want Celery to autoscale the number of concurrent processes, then we would probably want to dynamically scale the `qos` of the channel to match Celery.

xeonneon commented:
Changes since last post:
  - Removed the callback structure that was used for `basic_consume`
  - Use `basic_get` instead of `basic_consume` so we are in control of when the runner consumes a message from the queue
  - Slightly re-worked how often we query Celery for the Worker statuses

scott-taubman commented on runner/runner/listener.py:
```diff
@@ -59,6 +62,107 @@ def _handle_delivery(channel, deliver, properties, body):
+
+def _create_inspect() -> None:
+    """Create global Celery Inspect object to interact with Celery Worker API"""
+    global celery_inspect
```
I don't think there's a compelling reason to create a global here.  Other approaches you might take:

* Using `@cached` on `_get_inspect()` so that it only does the wait for connection once.
* Storing the result of `_get_inspect()` and just passing it into the `_wait_for_available_workers()` call:

```python
inspect = _get_inspect()

while True:
    method, properties, body = channel.basic_get("public")

    if method is None:
        sleep(WAIT_FOR_MESSAGE_DELAY)
        continue

    _handle_delivery(channel, method, properties, body)
    _wait_for_available_worker(inspect)
```

xeonneon commented on runner/runner/listener.py:
```diff
@@ -59,6 +62,107 @@ def _handle_delivery(channel, deliver, properties, body):
+    celery_inspect = _get_inspect()
+
+
+@cached(cache=TTL_CACHE)
```
Ok, makes sense. I'll need to slightly alter the code that checks if the worker is available so that it is just checking the worker's number of assigned tasks to that `WORKER_CONCURRENCY`, rather than a dictionary of the worker's concurrency levels. I'll also probably keep that `_get_worker_concurrency()` method just for the sake of unit testing.

xeonneon commented:
Changes:
  - Updated methods for checking for available worker processes
    - Only expects to work with one worker
    - Worker hostname is hardcoded for dictionary lookup versus popping first item from the dictionary
    - Since workers are isolated from each other due to vhosts, this should not be an issue
  - Updated unit tests to reflect this new method

Some methods are kind of unnecessary, but they help with being able to mock values in the unit tests, so I opted to leave them.
